RecoCTPPS/TotemRPLocal/plugins/TotemTimingLocalTrackFitter.cc

```diff
@@ -0,0 +1,164 @@
+
+  edm::Handle<edm::DetSetVector<TotemTimingRecHit> > recHits;
+  iEvent.getByToken( recHitsToken_, recHits );
+
+  for(const auto& trk_algo_entry: trk_algo_map)
+    pOut->find_or_insert(trk_algo_entry.first);
+
+  std::map< TotemTimingDetId, int > planeActivityMap;
+
+  for (const auto& vec: *recHits) {
+    const TotemTimingDetId detId(vec.detId());
+    TotemTimingDetId tmpId(0, 1, 0, 0, 0);
+    tmpId.setArm(detId.arm());
+    tmpId.setRP(detId.rp());
+    tmpId.setPlane(detId.plane());
+    planeActivityMap[tmpId] += vec.size();
+  }
+
+  // feed hits to the track producers
+  for ( const auto& vec : *recHits ) {
+
+    const TotemTimingDetId detId(vec.detId());
+
+    TotemTimingDetId tmpId(0, 1, 0, 0, 0);
+    tmpId.setArm(detId.arm());
+    tmpId.setRP(detId.rp());
+    tmpId.setPlane(detId.plane());
+    if (planeActivityMap[tmpId] > maxPlaneActiveChannels)
+      continue;
+
+    for ( const auto& hit : vec ) {
+
+      tmpId.setPlane(0);
+
+      if(trk_algo_map.find(tmpId) != trk_algo_map.end())
+        trk_algo_map.find(tmpId)->second.addHit(hit);
+
+      else
+        edm::LogWarning("TotemTimingLocalTrackFitter") << "Invalid detId for rechit: arm=" << detId.arm() << ", rp=" << detId.rp();
+    }
+  }
+
+  // retrieves tracks for all hit sets
+  for(auto& trk_algo_entry: trk_algo_map)
+    trk_algo_entry.second.produceTracks(pOut->operator[](trk_algo_entry.first));
+
+  iEvent.put( std::move( pOut ) );
+
+  // remove all hits from the track producers to prepare for the next event
+  for(auto& trk_algo_entry: trk_algo_map)
+    trk_algo_entry.second.clear();
+}
+
+void
+TotemTimingLocalTrackFitter::fillDescriptions( edm::ConfigurationDescriptions& descr )
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>( "recHitsTag", edm::InputTag( "totemTimingRecHits" ) )
+    ->setComment( "input rechits collection to retrieve" );
+  desc.add<int>( "verbosity", 0 )
+    ->setComment( "general verbosity of this module" );
+
+  desc.add<double>( "threshold", 1.5 )
+    ->setComment( "minimal number of rechits to be observed before launching the track recognition algorithm" );
+  desc.add<double>( "thresholdFromMaximum", 0.5 )
+    ->setComment( "threshold relative to hit profile function local maximum for determining the width of the track" );
+  desc.add<double>( "resolution", 0.01 /* mm */ )
+    ->setComment( "spatial resolution on the horizontal coordinate (in mm)" );
+  desc.add<double>( "sigma", 0. )
+    ->setComment( "pixel efficiency function parameter determining the smoothness of the step" );
+  desc.add<double>( "tolerance", 0. /* mm */)
+    ->setComment( "tolerance used for checking if the track contains certain hit" );
+  desc.add<int>( "maxPlaneActiveChannels", 3 /* mm */)
+    ->setComment( "threshold for discriminating noisy planes" );
+
+
+  desc.add<std::string>( "pixelEfficiencyFunction", "(TMath::Erf((x-[0]+0.5*[1])/([2]/4)+2)+1)*TMath::Erfc((x-[0]-0.5*[1])/([2]/4)-2)/4" )
```

Comment:
Use ([1]-0.05) instead of [1] in all the formulas to cope with border inefficiencies